modules/azure-auth-header/azure-auth-header.c

```diff
@@ -0,0 +1,226 @@
+
+struct _AzureAuthHeaderPlugin
+{
+  LogDriverPlugin super;
+
+  gsize   secret_len;
+  guchar *secret;
+
+  gchar   *workspace_id;
+  gchar   *method;
+  gchar   *path;
+  gchar   *content_type;
+};
+
+static gsize
+_get_rfc1123date(gchar *buf, gsize buf_len)
+{
+  time_t now = time(NULL);
+  struct tm gmt;
+
+  gmtime_r(&now, &gmt);
+
+  gchar format[] = "%a, %d %b %Y %H:%M:%S GMT";
+  gsize len = strftime(buf, buf_len, format, &gmt);
+
+  return len;
+}
+
+static void
+_get_rfc1123data_or_assert(gchar *date)
+{
+  gsize date_len = 0;
+  date_len = _get_rfc1123date(date, MAX_DATE_LEN);
+  g_assert(date_len);
+}
+
+static GString *
+_azure_auth_header_get_str_to_hash(AzureAuthHeaderPlugin *self, glong content_len, const gchar *date)
+{
+  gchar *fmt = "%s\n%ld\n%s\nx-ms-date:%s\n%s";
+  GString *str = g_string_new(NULL);
+
+  g_string_append_printf(str, fmt, self->method, content_len, self->content_type, date, self->path);
+
+  return str;
+}
+
+static guint
+_azure_auth_header_get_digest(AzureAuthHeaderPlugin *self, GString *input, guchar *digest)
+{
+  guint md_len = 0;
+
+  if (!HMAC(EVP_sha256(),
+            self->secret, self->secret_len,
+            (const guchar *)input->str, input->len,
+            digest, &md_len))
+    {
+      msg_error("Failed to generate Azure Auth Header HMAC",
+                evt_tag_str("str", input->str),
+                evt_tag_int("len", input->len));
+    }
+
+  return md_len;
+}
+
+static void
+_azure_auth_header_format_headers(AzureAuthHeaderPlugin *self, List *headers, const gchar *digest, const gchar *date)
+{
+  GString *auth_hdr = g_string_new(NULL);
+  GString *date_hdr = g_string_new(NULL);
+
+  g_string_printf(auth_hdr, "Authorization: SharedKey %s:%s", self->workspace_id, digest);
+  g_string_printf(date_hdr, "x-ms-date: %s", date);
+
+  list_append(headers, auth_hdr->str);
+  list_append(headers, date_hdr->str);
+
+  g_string_free(auth_hdr, TRUE);
+  g_string_free(date_hdr, TRUE);
+}
+
+static gboolean
+_append_headers(AzureAuthHeaderPlugin *self, List *headers, GString *body)
+{
+  g_return_val_if_fail(self->secret, FALSE);
+
+  gchar date[MAX_DATE_LEN] = {0};
+  _get_rfc1123data_or_assert(date);
+
+  GString *rawstr = _azure_auth_header_get_str_to_hash(self, body->len, date);
+
+  guchar digest[EVP_MAX_MD_SIZE] = {0};
+  gsize digest_len = _azure_auth_header_get_digest(self, rawstr, digest);
+  if (digest_len == 0)
+    {
+      g_string_free(rawstr, TRUE);
+      return FALSE;
+    }
+
+  gchar *digest_str = g_base64_encode(digest, digest_len);
+  _azure_auth_header_format_headers(self, headers, digest_str, date);
+  g_free(digest_str);
+
+  g_string_free(rawstr, TRUE);
+
+  return TRUE;
+}
+
+
+void
+azure_auth_header_secret_set_from_b64str(AzureAuthHeaderPlugin *self, const gchar *b64secret)
+{
+  g_free(self->secret);
+  self->secret = g_base64_decode(b64secret, &self->secret_len);
+
+  if (!self->secret)
+    msg_error("Failed to base64 decode Azure secret key");
```

Comment:
Most of the options can be validated only at run-time after trying to send the first message.

This can be pre-checked (at least if its a valid base64 string). Can we make this a bool function and fail with CHECK_ERROR() in the grammar?

Reply:
yes. I just did not touch some parts of the original code